joomla/joomla-cms
# [5.0] Fix ShowOnRule validation with wrong results

Pull Request for Issue #41788 .

### Summary of Changes

The PR currently fixes the issue linked above, but in addition it fixes a general conceptional error with the validation rule added by PR #41460 :
- Field names can not only contain alphanumeric characters but also dashes.
- Showon can be used with any kind of field, i.e. also e.g. text or editor fields, so the field values can contain any kind of characters, including the `:` or `!:` which separate the field name from the value in showon rules. This was completely broken with PR #41788 . This PR here fixes that.

Finally, it adds unit tests for the ShowOnRule validation rule, thanks @heelc29 for these.

### Testing Instructions

1. Go to Content -> Fields
2. Edit an existing field or create a new one
3. Go to Options tab
4. Edit the "Showon Attribute" field.
5. Try to set to one of the below values and save:
(empty value)
`box:value1`
`box::`
`box:!`
`box:3:21`
`box:#@\[] ` (the last one is a space)
`box!:value1`
`box!:`
`box:value1[OR]square:value1`
`box:value1[AND]square:value1`
`box:value1[OR]square!:value1`
`box:value1[AND]square!:value1`
`box:value1[AND]square:value1,value2`
`box:value1[AND]square!:value1,value2`
`box:value1[AND]square!:value1,value2[OR]square!:value1`
`box:value1,value2[AND]square:value1`
`box:value1[AND]square!:value1:value2`
`box-test-1:value1`
`box-test-1!:value1`
`box-test-1!:`
`box-test-1:3:21`
`box:value1[AND]square-test-1!:value1:value2`
`box:value1[OR]square-test-1:value1`
`box:value1[OR]square-test-1!:value1`
`box`
`[AND]box:value3[OR]square:2:3`
`[AND][OR]`
`box@abc:value1`
`box_abc:value1`
`box!abc:value1`
6. Try any other valid or invalid rules which come to your mind.

In addition, check the Drone logs for the unit tests without and with this PR, or run the unit tests yourself using `./libraries/vendor/bin/phpunit --testsuite Unit`, and compare the results.

### Actual result BEFORE applying this Pull Request

The following rules fail validation, but they should pass:
`box::`
`box:!`
`box:#@\[] ` (the last one is a space)
`box!:`
`box-test-1!:`

The following rules pass validation, but they should fail:
`[AND]box:value3[OR]square:2:3`
`box@abc:value1`
`box_abc:value1`
`box!abc:value1`

### Expected result AFTER applying this Pull Request

The following rules are validated with success:
(empty value)
`box:value1`
`box::`
`box:!`
`box:3:21`
`box:#@\[] ` (the last one is a space)
`box!:value1`
`box!:`
`box:value1[OR]square:value1`
`box:value1[AND]square:value1`
`box:value1[OR]square!:value1`
`box:value1[AND]square!:value1`
`box:value1[AND]square:value1,value2`
`box:value1[AND]square!:value1,value2`
`box:value1[AND]square!:value1,value2[OR]square!:value1`
`box:value1,value2[AND]square:value1`
`box:value1[AND]square!:value1:value2`
`box-test-1:value1`
`box-test-1!:value1`
`box-test-1!:`
`box-test-1:3:21`
`box:value1[AND]square-test-1!:value1:value2`
`box:value1[OR]square-test-1:value1`
`box:value1[OR]square-test-1!:value1`

The following rules fail validation:
`box`
`[AND]box:value3[OR]square:2:3`
`[AND][OR]`
`box@abc:value1`
`box_abc:value1`
`box!abc:value1`

The new unit tests run with success.

### Link to documentations
Please select:
